model/host/host.go

```diff
@@ -957,8 +975,74 @@ func (h *Host) setAwaitingJasperRestart(user string) error {
 	return nil
 }
 
-// MarkAsReprovisioning puts the host in a state that means it is going to be
-// reprovisioned.
+// SetNeedsReprovisionToNew marks a host that is currently using new
+// provisioning to provision again.
+func (h *Host) SetNeedsReprovisionToNew(user string) error {
+	// Ignore hosts that are not provisioned by us (e.g. hosts spawned by
+	// tasks) and legacy SSH hosts.
+	if utility.StringSliceContains([]string{distro.BootstrapMethodNone, distro.BootstrapMethodLegacySSH}, h.Distro.BootstrapSettings.Method) {
+		return nil
+	}
+
+	if err := h.setAwaitingReprovisionToNew(user); err != nil {
+		return err
+	}
+	if h.StartedBy == evergreen.User && !h.NeedsNewAgentMonitor {
+		return nil
+	}
+	return h.MarkAsReprovisioning()
+}
+
+func (h *Host) setAwaitingReprovisionToNew(user string) error {
+	// While these checks aren't strictly necessary since they're already
+	// filtered in the query, they do return more useful error messages.
+
+	switch h.NeedsReprovision {
+	case ReprovisionToNew:
+		return nil
+	case ReprovisionToLegacy:
+		return errors.New("cannot reprovision a host that is already converting to legacy provisioning")
```

Comment:
This is intentional. The button only reprovisions hosts that use new provisioning. Legacy provisioned hosts don't  shouldn't be reprovisioned because they are non-stateful (i.e. all they require is an evergreen binary on the host and don't need Jasper nor any of the other provisioning files).

Reply:
I mean, why don't we do the check the same way we did in `setAwaitingJasperRestart`?
```
	if h.NeedsReprovision != ReprovisionNone {
		return errors.Errorf("cannot reprovision when host is already reprovisioning")
	}
```

Reply:
Reprovisioning a host from scratch (e.g. converting a host to new provisioning) takes precedence over just restarting Jasper (which basically just does a bit of work to restart Jasper and that's it) because it includes all the Jasper restart logic and a bit more to ensure the host is fully provisioned.